datumaro/plugins/splitter.py

```diff
@@ -0,0 +1,478 @@
+            assert len(labels) == 1, \
+                "Expected exact one label for a DatasetItem"
+            ann = labels[0]
+            attributes = dict(ann.attributes.items())
+            assert attr_pid in attributes, \
+                "'%s' is expected as attribute name" % attr_pid
+            person_id = attributes[attr_pid]
+            if person_id not in by_pid:
+                by_pid[person_id] = []
+            by_pid[person_id].append((idx, ann))
+            groups.add(ann.group)
+
+        max_group_id = max(groups)
+        self._group_map["gallery"] = max_group_id + 1
+        self._group_map["query"] = max_group_id + 2
+
+        required = self._get_required(id_ratio)
+        if len(by_pid) < required:
+            log.warning("There's not enough IDs, which is {}, \
+                so train/val/test ratio can't be guaranteed." % len(by_pid))
+
+        ## 1. split dataset into trval and test
+        ##    IDs in test set should not exist in train/val set.
+        test = id_ratio[id_subsets.index("test")] if "test" in id_subsets else 0
+        if test > NEAR_ZERO: # has testset
+            split_ratio = np.array([test, 1.0-test])
+            person_ids = list(by_pid.keys())
+            np.random.shuffle(person_ids)
+            sections = self._get_sections(len(person_ids), split_ratio)
+            splits = np.array_split(person_ids, sections)
+            testset = { pid: by_pid[pid] for pid in splits[0] }
+            trval = { pid: by_pid[pid] for pid in splits[1] }
+
+            ## follow the ratio of datasetitems as possible.
+            ## naive heuristic: exchange the best item one by one.
+            expected_count = int(len(self._extractor) * split_ratio[0])
+            testset_total = int(np.sum([len(v) for v in testset.values()]))
+            self._rebalancing(testset, trval, expected_count, testset_total)
+        else:
+            testset = dict()
+            trval = by_pid
+
+        by_splits = dict()
+        for subset in self._subsets:
+            by_splits[subset] = []
+
+        ## 2. split 'test' into 'gallery' and 'query'
+        if len(testset)>0:
+            for person_id, items in testset.items():
+                indice = [idx for idx, _ in items]
+                by_splits['test'].extend(indice)
+
+            valid = ["gallery", "query"]
+            test_subsets, test_ratio = self._validate_splits(test_splits,valid)
+            by_groups = { s : [] for s in test_subsets }
+            self._split_by_attr(testset, test_subsets, test_ratio, by_groups,
+                                dataset_key=attr_pid)
+
+            # tag using group
+            for idx, item in enumerate(self._extractor):
+                for subset, split in by_groups.items():
+                    if idx in split:
+                        group_id = self._group_map[subset]
+                        item.annotations[0].group = group_id
+                        break
```

Comment:
Here, I use 'group' property to split 'test' subset into 'query' and 'gallery' set. 
It seems that the default subset doesn't support a hierarchy. 
In the original implementation, we add "Auxiliary" attribute to the annotation. 
Do you think I change the current implementation w.r.t the original one? 
Or do I have to implement the subset hierarchy? 
Please share your opinion.

Reply:
Yes, there in no exact support for split hierarchy yet. To do what you want, I'd create 2 extra subsets - `test_gallery` and `test_query`, this would look quite natural (at least, in Datumaro). Technically, subsets are extractors themselves, so it is possible to extend the implementation with some nesting. But it would require lots of changes in the importing/exporting code.

Reply:
Actually, the gallery and query sets are not subsets. They're a kind of tag. 
I need more time to resolve this, so let's leave this as it is at the moment.
I'll resolve this issue while applying the command line feature.